packages/grid/_modules_/grid/hooks/features/columns/useGridColumns.ts

```diff
@@ -43,11 +52,22 @@ export function useGridColumns(
     [props.columnTypes],
   );
 
+  /**
+   * If `initialState.columns.visibleColumnsModel` or `visibleColumnsModel` was defined during the 1st render, we are directly updating the model
+   * If not, we keep the old behavior and update the `GridColDef.hide` option (which will update the state model through the `GridColDef.hide` => `visibleColumnsModel` sync in `createColumnsState`
+   */
+  const shouldUseVisibleColumnModel = React.useRef(
```

Comment:
The basic idea here was to avoid switching between legacy and non legacy while the `hide` field exists because I fear it might create bugs.
If you control, I don't care about the `hide` field, so if you stop controlling and then update one `hide` field, you will regenerate the model from the columns `hide` field and thus loose the visibility status of all the other columns.

I can spend more time making sure every corner case is handled if you think the controlled => not controlled is something we can't do without in v5.

I feel like it is a good trade-off to implement this feature without breaking changes and then enable the controlled => not controlled pattern in v6.

Reply:
Ok, I didn't see that it was there to support the legacy API. I don't think it's worth to spend time trying to support uncontrolled => controlled. Just add a comment explaining that.